### PR TITLE
Add UNDIAL, a new unlearning algorithm

### PR DESCRIPTION
**Pull request by dong-river**
# What does this PR do?

- [x] Added UNDIAL, a new unlearning algorithm publised in [NAACL](https://arxiv.org/pdf/2402.10052).
- [x] Tested on TOFU

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Have you gone through the contributions [guide](../docs/contributing.md)?
- [x] Are your changes documented? Read documentation guidelines [here](../README.md#-further-documentation).

### CODE REVIEW

**Comment by molereddy** [comment]
Thank you for this!

Can you make these additions
1. There are methods which run on top of unlearning loss functions (like [SOUL](https://github.com/OPTML-Group/SOUL), GRU (https://github.com/locuslab/open-unlearning/pull/83). Since you are introducing a loss function, it would be great if you can make the `undial_loss` function available outside your trainer code so it can be imported in these methods as we begin to also support them.
2. We've added [guidelines](https://github.com/locuslab/open-unlearning/blob/main/docs/contributing.md) on contributing to our repository, including sharing experimental settings, results, scripts, and  to run. Could you review them and update this PR accordingly?

**Summary of steps to contribute**
- Create a folder under [community/methods/](https://github.com/locuslab/open-unlearning/tree/main/community/methods).
- Include your method details, results, hyperparameter search scripts, and steps.
- Update the [leaderboard](https://github.com/locuslab/open-unlearning/blob/main/community/leaderboard.md) with your results.

We've provided a template [here](https://github.com/locuslab/open-unlearning/tree/main/community/methods/template).

**Comment by dong-river** [comment]
@molereddy Thank you for the suggestions. I fixed the part I could understand. And send a private message about things I don't understand. I hope at least what I did in this PR is able to be merged. Let me know if there's any fatal issues. Thank you!

**Comment by molereddy** [comment]
I've merged our upstream changes onto your branch. If you can test the code and add some results, that would be great, else we can just test on our end and merge the code in in the coming days.

**Comment by dong-river** [comment]
Sorry for the delay. I had a bit trouble running my method on 7B which requires LoRA implementation as my method need to run TWO 7B model in parallel. It would be great if LoRA could be supported. 

I would really appreciate if you could test my methods like all other using the same standard and follow the hyperparam search in my description in the PR (which should not be hard given you have the pipeline ready). I also turned the PR for review.

**Review by Dornavineeth — APPROVED** [review]
Great Work.

Currently it is tested only on TOFU 1B model.
